### PR TITLE
0.0.10 - UI bugfixes

### PR DESCRIPTION
**Pull request by zathras-crypto**
WIP PR to contain further UI fixes.

### CODE REVIEW

**Comment by zathras-crypto** [comment]
Note, FYI I'm having a hard time finding more optimizations here - over the last couple days I've stepped through and timed almost everything (albeit only on linux) and nothing seems to take more than just a few ms at most nowadays (except for initial population of cached maps, but this only happens once at client startup).  

I still see freezes during blockchain sync though, so there must be something else in play (perhaps waiting for locks?) to cause it since we only do a refresh when we get blocks containing omni transactions and updating all of the UI components seems to take a couple hundred ms in total (assuming there are no changes needed to the displayed data)...

Long story short from my testing it seems we only expend a few hundred ms per omni-containing-block when syncing (and most blocks are not omni-containing) - shouldn't be enough to hang the client...

Anyway, just a heads up & i'll keep at it and try and nail this down :)

**Comment by dexX7** [comment]
The recent changes look fine!

I think it's okay for now. With a smaller wallet, the client was already responsive, and I assume it's even better now. :)

The tests with dummy entries showed the `QTableWidget` itself seems to be the cause for the major lags, and unfortunally there is probably no easy optimization (except no longer using the widget).

**Comment by zathras-crypto** [comment]
> The recent changes look fine!

Thanks dude :)

> The tests with dummy entries showed the QTableWidget itself seems to be the cause for the major lags, and unfortunally there is probably no easy optimization (except no longer using the widget).

This is what has me a little stumped - I completely agree when we pump the ```QTableWidget``` with items it slows things down, but when syncing the blockchain we get freezes taking in a couple hundred blocks that don't contain any transactions affecting the wallet (thus no changes to ```QTableWidget``` should occur)...  

A bit of fun can be had with ```--omniuiwalletscope=10``` hehe, it's blazing fast then :p !!!

**Comment by dexX7** [comment]
How/where did you see those freezes? While syncing with Qt? (not scanning?)

**Comment by zathras-crypto** [comment]
> How/where did you see those freezes? While syncing with Qt? (not scanning?)

I see them pretty much anytime sync with QT yep...  If you're using any kind of virtualization software, snapshot and then in a few days revert to that snapshot so you've got a few days to "catch up" - you'll probably see the QT UI freeze (window greys out and goes unresponsive) several times while it catches up.

It runs perfectly once synced - it's just while catching up I see the issue.

**Comment by dexX7** [comment]
I'm not sure, how to test this, but is it certain that these lags are related to the UI at all?

The parts that may play a role are transaction parsing and storing the file based persistence. On the other hand, but my data might be skewed somehow: while scanning/parsing on mainnet, the average time spent in `parseTransaction()` after the marker check was only around 1-1.4 ms.

**Comment by dexX7** [comment]
My build with this PR has bad data:

0.00000000 MSC is shown, actual balance is 0.08125000 MSC:
overview

Balance overview is empty (also when clicking on wallet totals):
totals

Send from is empty:
send

Correct balance is shown here:
mdex


I started only a few minutes ago. Could it be that initial data fetching/UI update is missing?

**Comment by zathras-crypto** [comment]
That'll be my next step I think, to time non-UI functions but I'm almost
certain it's the UI because the debug window is unresponsive, but run same
test with the daemon and RPC is fully responsive...

**Comment by dexX7** [comment]
> I started only a few minutes ago. Could it be that initial data fetching/UI update is missing?

After the last block, and after clicking on wallet totals in the balance overview, the balance popped up. It's zero on the overview though.

**Comment by zathras-crypto** [comment]
Oh yikes!

I think it might be related to my set_wallet_totals changes since that's
now called by the block handler only when Omni transactions are present - I
must have forgotten to add one at init as well...  Sorry bud :(  I'm taking
the missus out for a much deserved dinner now (she's been really great
through the health stuff so deserves a treat he he) but I'll take a look &
fix it up tomorrow :)

**Comment by dexX7** [comment]
Aah. :)

Have a good time!

**Comment by zathras-crypto** [comment]
Still trying to get to the bottom of this one FYI - it's quite a strange issue - I've debugged it down to the point I can see that ```IsMyAddress()``` actually returns false for keys in the wallet when ```set_wallet_totals()``` is called from ```mastercore_init``` either directly or via ```msc_initial_scan``` - there must be some kind of order of precedence I'm not seeing here between when we init and when the keystore for the wallet is actually available to query...  

Narrowing it down hehe :)

**Comment by zathras-crypto** [comment]
OK hopefully that should have sorted it mate :)

**Comment by dexX7** [comment]
Argh! :)

Due to the missing BTC balance signal, the fee warning doesn't vanish after coins were received.

**Comment by dexX7** [comment]
I've seen segfaults, when running the regtests + QT.

I think this is the first time we face locking issues in our own code, which is sort of a good thing, because it shows we're moving away from the single threaded bottleneck.

I believe it's due to the UI signals, but it's not entirely clear to me what happens, and it looks like there is more than one path to cause a segfault. Here is some information, including stacktraces and debug lines:

- caused by `PopulateBalances()`: http://pastebin.com/raw.php?i=3eBdnvAd
- caused by `PopulateBalances()`: http://pastebin.com/raw.php?i=n6UddeGT
- caused by `set_wallet_totals()`: http://pastebin.com/raw.php?i=i0w42PDX

You may try to reproduce the issue as follows:
```shell
# Start with a clean state:
$ rm -rf ./bitcoin/regtest
$ ./src/qt/bitcoin-qt

# In another terminal, run the tests:
$ git clone https://github.com/OmniLayer/OmniJ.git
$ cd ./OmniJ
$ ./gradlew :omnij-rpc:regTest

# Head over to the terminal with core dump (assuming there is one):
$ gdb ./src/qt/bitcoin-qt ./core.30609
(gdb) where                # for short overview
(gdb) bt full              # for detailed stack trace
```

**Comment by dexX7** [comment]
Oh my.. I have the bad feeling we need to start to use locks everywhere.

As far as I understand, `cs_tally` can be used to lock the globals in `mastercore.cpp`, which is a pretty wide scope, probably too wide actually, but anyway, I've added a lock to `PopulateBalances()` and `set_wallet_totals()`, only to see that other places, such as the MetaDEx dialogs etc. etc ...

But that aside, let's assume we start to use locks properly (we definitely should!), then I'm still wondering, if it's the right thing to do, to update all wallet balances like this.

The benchmarking on Windows showed that `set_wallet_totals()` is one of the biggest time consumers.

Just to provide some context, but please note this is with extensive locking:

- QT with set_wallet_totals():
  3 min 49 seconds + more than 1 minute UI freeze on top

- QT without set_wallet_totals():
  2 min 38 seconds

- Daemon with set_wallet_totals():
  2 min 33 seconds

- Daemon without set_wallet_totals():
  2 min 25 seconds

I've tested to use a signal to initiate the balance update, but unfortunally this had a side effect: the balances are updated asynchronously, and even after the tests finished, the updates kept pouring in.

This attempt can be found here:
- https://github.com/dexX7/bitcoin/compare/24004e09749c15c46ef367b0436c58ab29eddc31...dexX7:24efa944cf03f2a90264fbc9c424ec0d4d01d203

And with overly extensive use of locks (should not be done this way!), and a first step to move Omni Core related things into a new model:
- https://github.com/dexX7/bitcoin/compare/24004e09749c15c46ef367b0436c58ab29eddc31...dexX7:54c00ded3ff7455acf6f5468c318669feb4fc43d

**Comment by zathras-crypto** [comment]
> Argh! :)  Due to the missing BTC balance signal, the fee warning doesn't vanish after coins were received.

Shoot - good catch...  We may still need the wallet model for BTC change notifications in send and metadex dialogs - I'll take another look...

**Comment by zathras-crypto** [comment]
> The benchmarking on Windows showed that set_wallet_totals() is one of the biggest time consumers.

With a large wallet?  The larger the wallet the longer this takes I believe...

> Just to provide some context

So just to confirm, on the daemon ```set_wallet_totals``` does not actually add a large overhead, but on QT it does?

**Comment by dexX7** [comment]
> With a large wallet? The larger the wallet the longer this takes I believe...

Yeah, the benchmarking was done with a larger wallet (but only 3-4 properties owned). I didn't test the updated code though.

> So just to confirm, on the daemon set_wallet_totals does not actually add a large overhead, but on QT it does?

Correct.

I assume it's due to secondary effects: on the overviewpage, and sendmpdialog, the globals are used to initiate other actions, such as updating property balances or the property selector. And this code isn't executed in the non-UI runs.

**Comment by zathras-crypto** [comment]
> Yeah, the benchmarking was done with a larger wallet (but only 3-4 properties owned). I didn't test the updated code though.

I believe larger wallets hurt us all round at the moment...  FYI ```set_wallet_totals``` makes extensive use of ```IsMyAddress()``` which does a keystore lookup for each address in ```mp_tally_map``` - I haven't timed it but my assumption would be the larger the keystore the slower ```set_wallet_totals``` will be...

> I assume it's due to secondary effects: on the overviewpage, and sendmpdialog, the globals are used to initiate other actions

The globals shouldn't initiate anything - ```set_wallet_totals``` should now only run once per block and only then if there were Omni transactions within the block.  That shouldn't be any different between daemon and UI...

**Comment by dexX7** [comment]
> set_wallet_totals

During the RPC tests `set_wallet_totals` is 1549x called, and the average time spent during each call is 6.059 ms, and in total 9.383 s. This roughly matches the difference between non-UI run with, and without `set_wallet_totals()`.

> The globals shouldn't initiate anything

I was referring to the following parts:

- [OverviewPage::updateOmni()](https://github.com/zathras-crypto/omnicore/blob/52160c23aece02a209c5be3136f22b9bde42828d/src/qt/overviewpage.cpp#L285)
- [SendMPDialog::updatePropSelector()](https://github.com/zathras-crypto/omnicore/blob/52160c23aece02a209c5be3136f22b9bde42828d/src/qt/sendmpdialog.cpp#L91)
- [BalancesDialog::UpdatePropSelector()](https://github.com/zathras-crypto/omnicore/blob/52160c23aece02a209c5be3136f22b9bde42828d/src/qt/balancesdialog.cpp#L131)

If the globals are not filled, then these will have less work to do.

**Comment by zathras-crypto** [comment]
> If the globals are not filled, then these will have less work to do.

Ahh, that makes sense...

Also I noticed on another of your commits about ```set_wallet_totals``` only being applicable to the UI - I just wanted to note that it's *currently* only applicable to the UI, but I think it can also be beneficial on the RPC layer - just the same as how you can retrieve the wallet balance over RPC for bitcoin, at some point I'd like to drop in wallet balance calls for Omni into RPC.

**Comment by zathras-crypto** [comment]
> During the RPC tests set_wallet_totals is 1549x called, and the average time spent during each call is 6.059 ms, and in total 9.383 s.

Perhaps it's just me, but I don't see that as a major problem...

If ```set_wallet_totals``` is called 1500 odd times, that means we've processed some 1500 blocks containing Omni transactions - if that were on mainnet script verification would have eclipsed that 9.383s.  

I don't want to miss the point, but this says to me we're looking at 10 seconds of compute for 10 days worth of blocks (1440) assuming every single bitcoin block mined contains Omni data - forgive the Q but - is that an issue?

**Comment by dexX7** [comment]
> Perhaps it's just me, but I don't see that as a major problem...

Hehe, sorry. We're talking across a bit right now, I think. :)

I initially assumed `set_wallet_totals` directly causes the delays, and if we're going to update the balances for each pending transaction, too, then I assumed this would make the issue twice as worse (in the worst case). This doesn't of course mean it's not the right thing to do.

Now that it's clear the wallet balance updates play only a minor role, I'm a lot less concerned. (I time tested it only a few minutes ago.)

**Comment by zathras-crypto** [comment]
OK cool :)

I am concerned about the segfaults you were seeing though :(

**Comment by dexX7** [comment]
With the following changes the segfaults seem to be resolved:

- https://github.com/dexX7/bitcoin/compare/886fe553fe70a8aea569c94a78a81aecf9ed94be...dexX7:be772d8581d354b67e392920470399a75ffd3d4d

However, and this is a bit odd, the time to run the regtests took 3:34 min (about a min longer than non-UI), but 3:38 min *on top* until the UI freeze was over. I messured it by triggering a shutdown via `strg+c` (or `bitcoin-cli stop`), and then waited for the actual shutdown.

**Comment by zathras-crypto** [comment]
Heh, I was working on similar changes re the property list :)

**Comment by dexX7** [comment]
It looks like the time on top, where the client remains frozen, is spent to catch up with the updates. I'm currenty logging all relevant calls, and after the tests finished, `balancesUpdated`, `UpdateOffers`, `UpdateHistory`, ... were only half way through.

Here are some additional numbers:
```
Time for tests:        04:26 min
Time until shutdown:   07:58 min (+ 03:32 min)

BalancesDialog::balancesUpdated():      117.769 ms/update,  177.831728 s total for  1510 calls
SendMPDialog::balancesUpdated():         80.448 ms/update,  121.557502 s total for  1511 calls
TXHistoryDialog::UpdateHistory():        59.323 ms/update,   89.636633 s total for  1511 calls
TXHistoryDialog::UpdateConfirmations():   8.072 ms/update,   12.689337 s total for  1572 calls
set_wallet_totals():                      4.943 ms/update,    7.657179 s total for  1549 calls
MetaDExDialog::UpdateOffers():            1.144 ms/update,    1.728485 s total for  1511 calls
OverviewPage::UpdatePropertyBalance():    1.536 ms/update,    1.417680 s total for   923 calls
                                        --------------------------------------
                                                            412.518544 s (~ 06:43 min)
```

I'm surprised by the sum of these numbers.. probably due to parallel execution though.

**Comment by zathras-crypto** [comment]
I'm working on something I think should really, really help this whole process out and optimize significantly - will push soon :)

**Comment by zathras-crypto** [comment]
Thanks heaps for the numbers, it will be interesting to see how they come up after this latest optimization :)

**Comment by dexX7** [comment]
Hehe, now I'm really curious! :)

**Comment by zathras-crypto** [comment]
Let me know what you think :)

**Comment by dexX7** [comment]
Give me a few minutes to test it! :)

**Comment by dexX7** [comment]
Wooho, time doiwn to 02:41 min + 39 seconds freeze on top.

**Comment by zathras-crypto** [comment]
> Wooho, time doiwn to 02:41 min + 39 seconds freeze on top.

Is that compared to a or b mate?
> a) Time for tests:        04:26 min
> b) Time until shutdown:   07:58 min (+ 03:32 min)

**Comment by dexX7** [comment]
To compare:
```
Time for tests:        02:41 min
Time until shutdown:   03:20 min (+ 00:39 min)
```

In other words:

After 2:41 min the tests finished, but it took 39 seconds on top until the client was responsive and successfully shutdown, and in total that's 3:20 min. :)

The number from above is slightly higher due to heavy debug printing, but not very much, I think.

**Comment by zathras-crypto** [comment]
Alright, excellent :) thanks heaps for the timing info mate, I'm really glad it seems to be working out well :)

**Comment by dexX7** [comment]
Right now there are freezes, because the arrival rate of new jobs/signals fired is higher than the service/processing rate, or put in other words: the UI can't simply keep up. Once the queue clears, the freeze resolves. Is that correct so far?

I'm wondering: if we could drop new jobs, which are going to be superseded anyway by newer jobs, then we may prevent the forming of a queue right from the start.

A simple approach could be to set a flag to indicate "there is something currently being processed", and that flag could be cleared, once the processing finished. If the flag is set and a new job arrives, the job is ignored.

This is probably not sufficient, and it should be ensured that the latest state is up-to-date (and the latest update request wasn't dropped).

Would that work and make sense?

**Comment by zathras-crypto** [comment]
> I'm wondering: if we could drop new jobs, which are going to be superseded anyway by newer jobs, then we may prevent the forming of a queue right from the start.

That's something we could research sure - basically we're queuing up QT messages so if there is some way to say when we emit a signal "if this signal is already in the message queue, don't duplicate it" then that would certainly be a good step...

**Comment by dexX7** [comment]
An alternative to the init.cpp update:
```diff
diff --git a/src/init.cpp b/src/init.cpp
index f562ab9..1c06213 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -74,7 +74,6 @@ CClientUIInterface uiInterface;
 // Omni Core initialization and shutdown handlers
 int mastercore_init();
 int mastercore_shutdown();
-void CheckWalletUpdate();
 
 //////////////////////////////////////////////////////////////////////////////
 //
@@ -1288,10 +1287,6 @@ bool AppInit2(boost::thread_group& threadGroup)
                 }
             }
         }
-
-    // Omni Core code should be initialized and wallet should now be loaded, perform an initial populate of Omni balances and cache
-    CheckWalletUpdate();
-
     } // (!fDisableWallet)
 #else // ENABLE_WALLET
     LogPrintf("No wallet compiled in!\n");
diff --git a/src/qt/overviewpage.cpp b/src/qt/overviewpage.cpp
index bbe3876..7c7239d 100644
--- a/src/qt/overviewpage.cpp
+++ b/src/qt/overviewpage.cpp
@@ -397,6 +397,9 @@ void OverviewPage::setWalletModel(WalletModel *model)
 
         updateWatchOnlyLabels(model->haveWatchOnly());
         connect(model, SIGNAL(notifyWatchonlyChanged(bool)), this, SLOT(updateWatchOnlyLabels(bool)));
+
+        // Omni Core code should be initialized and wallet should now be loaded, perform an initial populate of Omni balances and cache
+        CheckWalletUpdate();
     }
 
     // update the display unit, to not use the default ("BTC")

```

**Comment by zathras-crypto** [comment]
I think the concept of wallet totals will be extended to the RPC layer (at the moment you can't ask RPC a simple question like "how much MSC do I have?") so that initial population probably needs to stay outside the UI :)

**Comment by dexX7** [comment]
> how much MSC do I have?

Like: "how much MSC do I have in total"?

I'm not entirely convinced, because a command like this would likely try to update the cache, too, instead of blindly returning some numbers. But I also don't object to the update via init.cpp for now.

**Comment by zathras-crypto** [comment]
That's a fair point - I guess I was going for keeping the balance controlled by the core and having UI & RPC consume, rather than trigger balance updates - but of course it's all up flexible right now :)

**Comment by dexX7** [comment]
> I guess I was going for keeping the balance controlled by the core and having UI & RPC consume, rather than trigger balance updates

Ah, that's very well put, and I like this way of thinking. :)

**Comment by dexX7** [comment]
FWIW an overview of the current event sytem:

 - https://gist.github.com/dexX7/61a0a405f17af3b70f14

**Comment by dexX7** [comment]
Just as quick remainder:

With `-debug=ui`:
```
UpdateTip: new best=36d95cfd2015586b6c21808e26d3f00f0eb4606e072d500ce26b3e..
AddToWallet 731c2eed0b0fcce7bdf00f88de0bb1b180838a31754fb4b06269290b6dfa1d94  new
keypool return 1182
GUI: QTableWidget: cannot insert an item that is already owned by another QTableWidget
GUI: QTableWidget: cannot insert an item that is already owned by another QTableWidget
...
AddToWallet d3d3ffc1c1ea1a54cb33c244f1092bf68e611d239ad47831244adc4051fb3705  new
```

**Comment by zathras-crypto** [comment]
Ahh yes, thanks dude I'll look into that in a bit - I'm just running some tests for P2SH for Craig because he mentioned some things had stopped working in 0.0.10 - I'll CC you in on my response to him mate to keep you in the loop :)

**Comment by dexX7** [comment]
> P2SH

That's interesting. Please forward everything you have! :)

**Comment by zathras-crypto** [comment]
> Just as quick remainder:

Did you happen to notice what was being updated?

**Comment by dexX7** [comment]
It's triggered once a block confirms, and I think it's related to the trade history.

**Comment by zathras-crypto** [comment]
Alrighty, thanks dude :)

**Comment by dexX7** [comment]
It's caused by cancel-trade transactions (and not by sends or trades).

**Comment by zathras-crypto** [comment]
Ahh that helps thanks, I was having trouble replicating :)

**Comment by dexX7** [comment]
It might not be written to the `debug.log` without `-debug=ui`, or maybe even `-debug=1`.

**Comment by zathras-crypto** [comment]
Note, I'm seeing some quite strange behaviour :( so I'll be giving the MetaDEx UI a bit of a going over tomorrow and hopefully clean it up a bit.

**Comment by dexX7** [comment]
> strange behaviour

What do you see?

**Comment by zathras-crypto** [comment]
> What do you see?

* Duplicate trades in history with bad data
* Missing trades
* Cancel trades showing as new sells
* Lack of updates to history when sending a cancel transaction
* Lack of pending support when sending a cancel transaction (likely cause of above)
* Duplicate parsing where viewing a trade parses several times

In a nutshell I think I must have thought the MetaDEx UI was in better shape than it actually is - none of the problems are major, just bugs that I'll need to hunt down and figure out, so I'm giving myself a dedicated day tomorrow to focus purely on MetaDEx UI (I bounce between too many different things at the moment) and give it a good clean up :)

**Comment by dexX7** [comment]
Urgh.. :)

**Comment by zathras-crypto** [comment]
Yep :)

**Comment by zathras-crypto** [comment]
I know there is still work to be done on the "minimum fee warning", but the fee in general seems to be quite a bit higher than necessary:
screenshot from 2015-06-24 14 02 39

**Comment by dexX7** [comment]
I'll take a look at the fees after I finished the SP serialization. :)

Are those amounts shown in the picture the fees?

Which settings do you have in Send - Bitcoin - Transaction fee?

**Comment by zathras-crypto** [comment]
> I'll take a look at the fees after I finished the SP serialization. :)

Sure sure dude, just wanted to note it so I didn't forget about it hehe

> Are those amounts shown in the picture the fees?

**Comment by dexX7** [comment]
> Left at defaults because that's how end users are likely to use it - but seems unusually high 

Hm...

> the fee in general seems to be quite a bit higher than necessary

As far as I understand, with this setting the fee selection is up to the wallet.

Let me rephrase: do you think the fees are unexpected and too high, or do you think the wallet fee selection should be smarter?

**Comment by zathras-crypto** [comment]
> do you think the fees are unexpected and too high, or do you think the wallet fee selection should be smarter?

Well, perhaps I've misunderstood things but 6K to 10K satoshi mining fee seems well over the top, when an average transaction used to be a flat 1K?

**Comment by zathras-crypto** [comment]
P.S. even with a 6K fee on my latest test tx (72baa1f2af9886d810b679a2d8e4b5928f89804b43f8776d0d365d1a3eaf16f7) it's been unconfirmed for 10 blocks already hehe

**Comment by dexX7** [comment]
0.00006060 BTC isn't very high imho.

Via the RPC command `estimatefee n` the fee estimation can be retrieved, where n is the confirm target in blocks, and the value is per 1000 byte. If the result of `estimatefee` roughly matches the fees you paid, then it's indeed the auto fee selection by the wallet.

**Comment by zathras-crypto** [comment]
Not crazy high nope, but does highlight the other approach for the minimum warning is invalid (as that's always works out to 2600 satoshis, which is not sufficient to send).  Will loop back around to the fee issue...

**Comment by dexX7** [comment]
Hey @zathras-crypto, I would like to start merging the outstanding PRs. Could you please signal, if there is anything that should be held back? In particular I was thinking about this one + #74, ideally #88, too, to increase the exposure.

**Comment by zathras-crypto** [comment]
> Hey @zathras-crypto, I would like to start merging the outstanding PRs. Could you please signal, if there is anything that should be held back? In particular I was thinking about this one + #74, ideally #88, too, to increase the exposure.

Hey bud, yeah we need to merge everything in because @CraigSellars wants release candidate binaries basically yesterday hehe...

Let me step through these outstanding ones again and see where we're at :)

**Comment by dexX7** [comment]
Hey there!

> we need to merge everything in because @CraigSellars wants release candidate binaries basically yesterday hehe...

This is going to be interesting, hehe. Is he looking for a release candidate or a preview? What's the target audience?

**Comment by zathras-crypto** [comment]
> Is he looking for a release candidate or a preview? What's the target audience?

Something people can get their hands on and start playing with for testing purposes.  @CraigSellars can you advise on target audience?

**Comment by dexX7** [comment]
Alright. I'm currently building dependencies via Gitian, but I won't wait until it's finished (it takes hours.. usually hehe).

So just to confirm: all PRs are good to go, in your opinion?

**Comment by zathras-crypto** [comment]
> So just to confirm: all PRs are good to go, in your opinion?

I wouldn't say that mate - I'm sure they'll be fine and 0.0.10 is still development but I'd like to review - I'm just working regtest and this UI reorg thing and then I'll switch gears and start reviewing your outstanding PRs mate (which I'm sure will be stellar :) )

**Comment by dexX7** [comment]
Hehe, thanks. We'll see. I think there might be one or two things you won't like (me neither! -> "prefix"). :)

Just let me know, I'm very curious how it all fits together.

**Comment by zathras-crypto** [comment]
This regtest stuff is definitely easier than I thought haha - can't believe I procrastinated so long!

First issue I'm coming up against is after mining 101 blocks and then sending some BTC to moneyqman and then mining more blocks, I don't seem to get any MSC generated...  Looking into it now...

**Comment by dexX7** [comment]
IIRC Exodus purchases don't go well with the UI, for some reason. It helps to check the log, and use the RPC console to get more information.

If all this doesn't help, then you could quickly edit your `bitcoin.conf`
```
server=1
rpcuser=bitcoinrpc
rpcpassword=pass
rpcallowip=127.0.0.1
txindex=1
```
And start QT. Then, from anywhere:
```
git clone https://github.com/OmniLayer/OmniJ.git
cd OmniJ
./gradlew :omnij-rpc:regTest
```
... and stop inbetween, so your wallet has some MSC. It might be necessary to mine a block afterwards via `setgenerate true 1`.

The issue I initially described can be reproduced by creating a bunch of Omni transactions, and then invaliding a very early block, say:
```
./src/bitcoin-cli getblockhash 1
...
./src/bitcoin-cli invalidateblock [hash]
./src/bitcoin-cli setgenerate true 1
```

**Comment by zathras-crypto** [comment]
I whipped up this quick script:
```
cat regtest.sh 
#!/bin/sh
src/bitcoin-cli --regtest setgenerate true 101
src/bitcoin-cli --regtest getnewaddress
echo "Please enter the new address: "
read addressvar
src/bitcoin-cli --regtest sendtoaddress $addressvar 10
src/bitcoin-cli --regtest setgenerate true 1
src/bitcoin-cli --regtest setaccount $addressvar "TESTACCOUNT"
src/bitcoin-cli --regtest sendfrom "TESTACCOUNT" moneyqMan7uh8FqdCA2BV5yZ8qVrc9ikLP 9
src/bitcoin-cli --regtest setgenerate true 1
src/bitcoin-cli --regtest getbalance_MP $addressvar 1
```
Which should basically mine some coins, send then to a new address, then send from that new address to moneyman and generate some MSC.

```
[
...bunch of blocks...
]
mkdCV1MB6fkxNNi9zr7FxmYV9ubQmbBVCr
Please enter the new address: 
mkdCV1MB6fkxNNi9zr7FxmYV9ubQmbBVCr
94ca0eeab5b6c2d485e6b94c3a7f35c9a6617afad58d69f8a096336e35c11560
[
    "16fd31e14940abc474b6f92364860ed7dd6e06928841dfb4563602000e7cd6ed"
]
3c9f270100f7248805fa9e8b0c3bed184d672c72b2dfbcb47f2e5d8d7c2a635f
[
    "26b2364061f7ea73524cae1946d8b220a4241f3e32466c64ca218c5a13b6066f"
]
{
    "balance" : "900.00000000",
    "reserved" : "0.00000000"
}
```

So it seems to be working now - not sure what the issue was earlier.  

> The issue I initially described can be reproduced by creating a bunch of Omni transactions, and then invaliding a very early block

Thanks dude, I'm just spinning up a test scenario now so I can test properly myself :)

**Comment by zathras-crypto** [comment]
P.S. this is uber cool :)  I always thought I'd lose a whole bunch of productivity deviating away to play with regtest but now I think I'm going to see a whole lot of benefit from it :)

**Comment by dexX7** [comment]
> not sure what the issue was earlier

It could also be that coinbase coins were used earlier, which don't count.

Anyway, have fun! It's probably going to make you *much* more productive. :)

**Comment by zathras-crypto** [comment]
I think this can probably be merged in now to get consolidation moving...

**Comment by dexX7** [comment]
Awesome!  :)

Quick note, it doesn't pass the RPC tests, when run locally on my machine. With lock debugging:
```
POTENTIAL DEADLOCK DETECTED
Previous lock order was:
 (1) cs_tally  qt/metadexdialog.cpp:387
 (2) cs_main  omnicore/omnicore.cpp:160
Current lock order is:
 (2) cs_main  main.cpp:3123
 (2) cs_main  main.cpp:2213
 (1) cs_tally  omnicore/omnicore.cpp:308
AddToWallet eddcca1353d001444ecf480ec48703a395480205e976ac516c55277648c04800  new
```
I'll take a look at it now.

**Comment by dexX7** [comment]
@zathras-crypto:
Would it be complicated to attach the hourglass for pending transactions to the left? (... and make sense for you?)

When running the tests, it jumps back and forth the whole time, moving the other icons on the right around.

**Comment by zathras-crypto** [comment]
>@zathras-crypto:
Would it be complicated to attach the hourglass for pending transactions to the left? (... and make sense for you?)
When running the tests, it jumps back and forth the whole time, moving the other icons on the right around.

Hmm, to the left of the status bar or to the left of the other icons?  Shouldn't be too much trouble nope :)

**Comment by zathras-crypto** [comment]
You can try this, but I don't think it looks "right"
```
diff --git a/src/qt/bitcoingui.cpp b/src/qt/bitcoingui.cpp
index 03a7860..6395171 100644
--- a/src/qt/bitcoingui.cpp
+++ b/src/qt/bitcoingui.cpp
@@ -181,6 +181,8 @@ BitcoinGUI::BitcoinGUI(const NetworkStyle *networkStyle, QWidget *parent) :
     if(enableWallet)
     {
         frameBlocksLayout->addStretch();
+        frameBlocksLayout->addWidget(labelOmniPendingIcon);
+        frameBlocksLayout->addStretch();
         frameBlocksLayout->addWidget(unitDisplayControl);
         frameBlocksLayout->addStretch();
         frameBlocksLayout->addWidget(labelEncryptionIcon);
@@ -190,8 +192,6 @@ BitcoinGUI::BitcoinGUI(const NetworkStyle *networkStyle, QWidget *parent) :
     frameBlocksLayout->addStretch();
     frameBlocksLayout->addWidget(labelBlocksIcon);
     frameBlocksLayout->addStretch();
-    frameBlocksLayout->addWidget(labelOmniPendingIcon);
-    frameBlocksLayout->addStretch();
 
     // Progress bar and label for blocks download
     progressBarLabel = new QLabel();
```

**Comment by zathras-crypto** [comment]
P.S. how do you paste your diffs in colour?

**Comment by dexX7** [comment]
Ah, let me try this in a few minutes.

The highlighting:
```diff
-    deleted
+    added
```
Can be done by adding "diff" to "```diff":
```
'''diff
-    deleted
+    added
'''
```
Where `'` is actually: `

https://help.github.com/articles/github-flavored-markdown/#syntax-highlighting

**Comment by zathras-crypto** [comment]
> The highlighting

Brilliant, thanks for the tip :)

**Comment by dexX7** [comment]
Hmm.. so you don't like the second one with the text?

I kinda like it, especially with the text it's very visible and good to notice.

**Comment by zathras-crypto** [comment]
> Hmm.. so you don't like the second one with the text?
I kinda like it, especially with the text it's very visible and good to notice.

The first one I didn't like (to the left of icons) so I took a stab at a different approach which was the second bit - I actually kind of like it too :)  I'll commit it then :)

**Comment by dexX7** [comment]
Nice!

> You have Omni transactions pending confirmation

This sounds slightly twisted to be, due to the "pending confirmation", because I usually refer to "pending transactions". Is that wrong?

**Comment by zathras-crypto** [comment]
The tooltip is slightly different ("awaiting confirmation" rather than "pending confirmation", do you think that's better wording?

Perhaps we should include 'sent' or 'outbound' or similar?  Do any of these sound right to you?
* "You have Omni transactions awaiting confirmation."
* "You have Omni transactions pending confirmation."
* "You have recently sent Omni transactions that are awaiting confirmation."
* "Your sent Omni transactions are awaiting confirmation."
* "Sent Omni transactions are awaiting confirmation."

Any other suggestions, feel free bud :)

**Comment by dexX7** [comment]
The first two sound best:

- "You have Omni transactions awaiting confirmation."
- "You have Omni transactions pending confirmation."

So let's go with the first, I'd say?

**Comment by zathras-crypto** [comment]
> So let's go with the first, I'd say?

Sure, agreed :)
